src/Common/src/System/Security/Cryptography/PasswordBasedEncryption.cs

```diff
@@ -0,0 +1,1032 @@
+            }
+
+            AsnReader reader = new AsnReader(encryptionSchemeParameters.Value, AsnEncodingRules.BER);
+
+            if (!reader.TryCopyOctetStringBytes(iv, out int bytesWritten) || bytesWritten != length)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            reader.ThrowIfNotEmpty();
+            iv = iv.Slice(0, bytesWritten);
+        }
+
+        private static unsafe Rfc2898DeriveBytes OpenPbkdf2(
+            ReadOnlySpan<byte> password,
+            ReadOnlyMemory<byte>? parameters,
+            out byte? requestedKeyLength)
+        {
+            if (!parameters.HasValue)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            Pbkdf2Params pbkdf2Params =
+                AsnSerializer.Deserialize<Pbkdf2Params>(parameters.Value, AsnEncodingRules.BER);
+
+            // No OtherSource is defined in RFC 2898 or RFC 8018, so whatever
+            // algorithm was requested isn't one we know.
+            if (pbkdf2Params.Salt.OtherSource != null)
+            {
+                throw new CryptographicException(
+                    SR.Format(
+                        SR.Cryptography_UnknownAlgorithmIdentifier,
+                        pbkdf2Params.Salt.OtherSource.Value.Algorithm));
+            }
+
+            if (pbkdf2Params.Salt.Specified == null)
+            {
+                Debug.Fail($"No Specified Salt value is present, indicating a new choice was unhandled");
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            HashAlgorithmName prf;
+
+            switch (pbkdf2Params.Prf.Algorithm.Value)
+            {
+                case Oids.HmacWithSha1:
+                    prf = HashAlgorithmName.SHA1;
+                    break;
+                case Oids.HmacWithSha256:
+                    prf = HashAlgorithmName.SHA256;
+                    break;
+                case Oids.HmacWithSha384:
+                    prf = HashAlgorithmName.SHA384;
+                    break;
+                case Oids.HmacWithSha512:
+                    prf = HashAlgorithmName.SHA512;
+                    break;
+                default:
+                    throw new CryptographicException(
+                        SR.Format(
+                            SR.Cryptography_UnknownAlgorithmIdentifier,
+                            pbkdf2Params.Prf.Algorithm));
+            }
+
+            // All of the PRFs that we know about have NULL parameters, so check that now that we know
+            // it's not just that we don't know the algorithm.
+
+            if (!pbkdf2Params.Prf.HasNullEquivalentParameters())
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            int iterationCount = ClampIterationCount(pbkdf2Params.IterationCount);
```

Comment:
Why clamping? Will it work with different implementation? Is this defined by spec or limitation on our side (if latter should we rather throw?)

Reply:
It does throw.  So I'll rename it to ValidateIterationCount.